services/traction/api/jobs/job_list.py

```diff
@@ -0,0 +1,192 @@
+        for dependency in dependencies:
+            j = dependency(self.profile)
+            if j.status == TenantJobStatusType.active:
+                dependencies_active += 1
+            else:
+                await j.start()
+
+        if dependencies_active == len(dependencies):
+            # all dependencies completed... let's get started
+            await self._do_start()
+
+        self._logger.info(f"< start({dependencies})")
+
+    async def fire_event(self, job_event_type: TenantJobStatusType, payload: dict):
+        topic = f"{self.job_name()}::{job_event_type}".lower()
+        self._logger.info(f"firing event (topic={topic}, payload={payload})")
+        await self.profile.notify(
+            TRACTION_JOB_PREFIX, {"topic": topic, "payload": payload}
+        )
+
+    async def approve(self):
+        # most likely called by the innkeeper...
+        # just mark as approved, let other process start it
+        job = await self._get_job()
+        await TenantJob.update_by_id(
+            job.tenant_job_id, {"status": TenantJobStatusType.approved}
+        )
+
+    async def execute(self, topic: str, payload: dict):
+        # called by the job list... tell job to do something to handle event
+        self._logger.info(f"> execute({topic}, {payload})")
+        tenant = await self._get_tenant()
+        self._logger.info(f"tenant = {tenant}")
+        context["TENANT_WALLET_TOKEN"] = tenant.wallet_token
+        # fetch the current state of the job from db...
+        names = topic.split("::")
+        job_name = names[0]
+        event_name = names[1]
+        # try a specific function first (to handle event from other job)
+        on_func = f"on_{job_name}_{event_name}"
+        await self._call_event_func(on_func, payload)
+        if job_name == self.job_name():
+            # call our event handler...
+            on_func = f"on_{event_name}"
+            await self._call_event_func(on_func, payload)
+
+        self._logger.info(f"< execute({payload})")
+
+    async def _call_event_func(self, func_name: str, payload: dict):
+        if hasattr(self, func_name) and callable(func := getattr(self, func_name)):
+            await func(payload)
+
+    @classmethod
+    def job_name(cls):
+        return re.sub(r"(?<!^)(?=[A-Z])", "_", cls.__name__).lower()
+
+    async def status(self):
+        job = await self._get_job()
+        return job.status
+
+    @abstractmethod
+    async def _do_start(self):
+        pass
+
+    @abstractmethod
+    async def on_requested(self, payload: dict):
+        pass
+
+    @abstractmethod
+    async def on_approved(self, payload: dict):
+        pass
+
+    @abstractmethod
+    async def on_denied(self, payload: dict):
+        pass
+
+    @abstractmethod
+    async def on_processing(self, payload: dict):
+        pass
+
+    @abstractmethod
+    async def on_completed(self, payload: dict):
+        pass
+
+    @abstractmethod
+    async def on_error(self, payload: dict):
+        pass
+
```

Comment:
sounds great, whip up an example and let's do it!

Reply:
`state_map = {e.value: lambda: None for e in self.state_enum}` should be in the baseClass.__init__ and will instantiate a map of all the enum values to an empty function. 

then  

`self.state_map[self.state_enum(payload["state"])]()` will call the appropriate function. 

we probably want a fixed set of parameters for all of these functions, but we can set that in the lambda: None default like `lambda profile,payload:None`

Reply:
thanks. I'll give it a shot. all those methods have the same signature, so that should be ok. not sure how that will address handling subscriptions to other jobs states?

Reply:
We also want async, so ProtocolBase might just need to define the function properly instead of using a lambda. 
```
def async _empty_function(self, payload):
  pass
```

Reply:
Protocol is also a new type in python that is preferred over ABC in alot of cases, but I haven't used them myself, and haven't compared it to our usage of ABC's. 

https://www.pythontutorial.net/python-oop/python-protocol/

Reply:
looks like it will be useful for us.

Reply:
A very 'untested' example might look like this. I'll see if i can actually plug it in. 

```
from logging import Logger, getLogger
from enum import Enum
from typing import Callable, Dict, Pattern, Protocol

from api.endpoints.models.credentials import CredentialRoleType, CredentialStateType

from api.core.event_bus import Event
from api.core.profile import Profile
from api.endpoints.models.webhooks import WEBHOOK_ISSUE_LISTENER_PATTERN


class ProtocolBase(Protocol):
    state_enum: Enum = Enum()
    state_map: dict[state_enum, Callable]
    logger: Logger
    webhook_listener_pattern: Pattern

    async def _empty_handler(self, payload: dict):
        pass

    def __init__(self):
        self.logger = getLogger(type(self).__name__)
        self.state_map = {e.value: self._empty_handler for e in self.state_enum}

    async def notify(self, profile: Profile, event: Event):
        self.logger.info("> notify()")
        payload = event.payload["payload"]
        self.logger.debug(f"payload={payload}")

        await self.state_map[self.state_enum(payload["state"])]()

        self.logger.info("> notify()")


class IssueCredentialProtocol(ProtocolBase):
    state_enum = CredentialStateType
    webhook_listener_pattern = WEBHOOK_ISSUE_LISTENER_PATTERN

    def on_ack(self, payload: dict):
        self.logger.info("ISSUE CREDENTIAL PROTOCOL COMPLETED")
        pass

    def __init__(self):
        self = super(IssueCredentialProtocol, self.__class__)
        self.state_map[CredentialStateType.credential_acked] = self.on_ack

```

Reply:
added chunk of code to reduce the Job class and its children... not exactly the same but close enough to prove it out. very cool pattern and we should leverage it in future refactor. obviously, a huge chunk of that is getting the Enums squared away which is in the backlog tickets.